packages/graalvm/src/main/kotlin/elide/runtime/intrinsics/js/s3/S3Module.kt

```diff
@@ -0,0 +1,448 @@
+      ?: throw JsError.typeError("No options passed!")
+    val config = S3ClientConstructorOptions.from(options)
+    S3ClientProxy(this.guestExec, config)
+  }
+
+  override fun install(bindings: GuestIntrinsic.MutableIntrinsicBindings) {
+    ModuleRegistry.deferred(ModuleInfo.of(S3_MODULE_NAME)) {
+      ProxyObject.fromMap(mapOf(
+        S3_CLIENT_SYMBOL to constructor,
+      ))
+    }
+  }
+}
+
+public class S3ClientProxy (private val guestExec: GuestExecutorProvider,
+                            private val config: S3ClientConstructorOptions) : S3Client {
+  private val s3Client: S3AsyncClient
+  private val credentialsProvider: AwsCredentialsProvider
+  private val clientRegion: Region?
+
+  // This map represents the JavaScript object representation of the class to the guest environment
+  // i.e. it is a map from strings to methods (in this case)
+  private val jsObj: Map<String, Any> = mapOf(
+    "file" to ProxyExecutable { args ->
+      val path = args.getOrNull(0)?.asString() ?:
+        throw JsError.typeError("S3Client.file() requires one string argument")
+      file(path)
+    }
+  )
+
+  init {
+    val credentials = AwsBasicCredentials.create(config.accessKeyId, config.secretAccessKey)
+    this.credentialsProvider = StaticCredentialsProvider.create(credentials)
+    this.clientRegion = when {
+      config.region != null -> Region.of(config.region)
+      config.endpoint == null -> Region.US_EAST_1
+      else -> null
+    }
+
+    this.s3Client = S3AsyncClient.builder()
+      .credentialsProvider(credentialsProvider)
+      .apply {
+        config.endpoint?.let {
+          endpointOverride(URI.create(it))
+        }
+
+        clientRegion?.let {
+          region(clientRegion)
+        }
+
+        if (config.virtualHostedStyle != true) {
+          forcePathStyle(true)
+        }
+      }
+      .build()
+  }
+  @Polyglot override fun file(path: String): S3File {
+    return S3FileProxy(guestExec, s3Client, clientRegion, credentialsProvider, path, config.bucket, config.acl)
+  }
+
+  override fun getMember(key: String?): Any? = jsObj[key]
+  override fun getMemberKeys(): Any? = ProxyArray.fromArray(jsObj.keys)
+  override fun hasMember(key: String?): Boolean = jsObj.containsKey(key)
+  override fun putMember(key: String?, value: Value?): Nothing =
+    throw JsError.typeError("Cannot set property '$key' on S3Client object.")
+}
+
+public class S3FileProxy(
+  private val guestExec: GuestExecutorProvider,
+  private val s3Client: S3AsyncClient,
+  private val region: Region?,
+  private val credentialsProvider: AwsCredentialsProvider,
+  private val path: String,
+  private val bucket: String?,
+  private val acl: String?) : S3File {
+
+  // This map represents the JavaScript object representation of the class to the guest environment
+  // i.e. it is a map from strings to methods (in this case)
+  private val jsObj: Map<String, Any> = mapOf(
+    "write" to ProxyExecutable { args ->
```

Comment:
this is good, but even better:
```kotlin
private val jsObj: Map<String, Any> = buildMap {
  this["write"] = ProxyExecutable { ... }
}
```
using [`buildMap`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/build-map.html) allows you to establish a new block, which is more flexible than an arg suite

anyway you don't need to update this particular code, it's just good to know for later 👍